### PR TITLE
feat: handle eligible deal stats via spark-api

### PR DESCRIPTION
**Pull request by bajtos**
Rework the endpoints `{miner/client/allocator}/:id/deals/eligible/summary` to return a (temporary) redirect to spark-api, as discussed in https://github.com/filecoin-station/spark-stats/pull/194#discussion_r1713500083.

Links:
- https://github.com/filecoin-station/spark-api/pull/403
- https://github.com/filecoin-station/spark/issues/78

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (juliangruber) [inline .github/workflows/ci.yml:1]: 😍

File: stats/lib/handler.js [hunk 1]
```diff
@@ -141,86 +143,18 @@ const notFound = (res) => {
...
-    WHERE client_id = $1 AND expires_at > now()
-    GROUP BY miner_id
-    ORDER BY deal_count DESC, miner_id ASC
-    `, [
-    clientId
-  ])
-
+const redirectToSparkApi = async (req, res, SPARK_API_BASE_URL) => {
+  console.log('SPARK_API_BASE_URL', SPARK_API_BASE_URL)
+  console.log('req.url', req.url)
```
Comment (juliangruber) [inline stats/lib/handler.js:152]:
```suggestion
```

File: stats/lib/handler.js [hunk 2]
```diff
@@ -141,86 +143,18 @@ const notFound = (res) => {
...
-    clients: rows.map(
-      // eslint-disable-next-line camelcase
-      ({ client_id, deal_count }) => ({ clientId: client_id, dealCount: deal_count })
-    )
-  }
-  json(res, body)
+  const location = new URL(req.url, SPARK_API_BASE_URL).toString()
+  res.setHeader('location', location)
+  res.statusCode = 302
+  res.end(location)
```
Comment (juliangruber) [inline stats/lib/handler.js:159]:
```suggestion
  redirect(req, res, location, 302)
```

https://github.com/juliangruber/http-responders/blob/5820864da9cb4728d4d86e76184eb554a92c16ed/index.js#L15

File: stats/lib/handler.js [hunk 3]
```diff
@@ -141,86 +143,18 @@ const notFound = (res) => {
...
-    ORDER BY deal_count DESC, miner_id ASC
-    `, [
-    clientId
-  ])
-
+const redirectToSparkApi = async (req, res, SPARK_API_BASE_URL) => {
+  console.log('SPARK_API_BASE_URL', SPARK_API_BASE_URL)
+  console.log('req.url', req.url)
   // Cache the response for 6 hours
   res.setHeader('cache-control', `max-age=${6 * 3600}`)
```
Comment (juliangruber) [inline stats/lib/handler.js:154]: I believe this response can be cached for a lot longer, but also that it doesn't really make a big difference

File: stats/bin/spark-stats.js [hunk 4]
```diff
@@ -17,7 +18,7 @@ const logger = {
   request: ['1', 'true'].includes(REQUEST_LOGGING) ? console.info : () => {}
 }
 
-const handler = createHandler({ pgPools, logger })
+const handler = createHandler({ SPARK_API_BASE_URL, pgPools, logger })
```
Comment (juliangruber) [inline stats/bin/spark-stats.js:21]: in this PR, `SPARK_API_BASE_URL` is never overridden, so it should be removed from args and added to config instead

**Review by juliangruber — CHANGES_REQUESTED** [review]

**Comment by juliangruber** [comment]
@bajtos there's a type error

File: stats/lib/handler.js (hunk 1 above)
Comment (bajtos) [reply stats/lib/handler.js:152]: 🙈

File: stats/bin/spark-stats.js (hunk 4 above)
Comment (bajtos) [reply stats/bin/spark-stats.js:21]: I want to make this configurable via env vars so we can run spark-stats locally with redirects to the local spark-api instance. With your proposal, I would have to edit the local config file and then remember to discard the changes before committing to git.

In that light, do you still prefer to move this to the config file? I don't have a strong opinion.

Comment (juliangruber) [reply stats/bin/spark-stats.js:21]: Got it, I didn't see that possibility. I never run these services in dev 😅

**Comment by bajtos** [comment]
> @bajtos there's a type error

Do you know why we don't run `npm test:types` as part of `npm test`? Was it an intentional decision? If not, then I'd like to fix that.

File: stats/lib/handler.js (hunk 3 above)
Comment (bajtos) [reply stats/lib/handler.js:153]: 💯

**Comment by bajtos** [comment]
All CI checks are green now 🚀

**Comment by juliangruber** [comment]
> > @bajtos there's a type error
> 
> Do you know why we don't run `npm test:types` as part of `npm test`? Was it an intentional decision? If not, then I'd like to fix that.

Yes please add that 👌

File: stats/lib/handler.js [hunk 5]
```diff
@@ -141,86 +144,14 @@ const notFound = (res) => {
...
-  const { rows } = await client.query(`
-    SELECT miner_id, COUNT(cid)::INTEGER as deal_count FROM retrievable_deals
-    WHERE client_id = $1 AND expires_at > now()
-    GROUP BY miner_id
-    ORDER BY deal_count DESC, miner_id ASC
-    `, [
-    clientId
-  ])
-
+const redirectToSparkApi = async (req, res, SPARK_API_BASE_URL) => {
```
Comment (juliangruber) [inline stats/lib/handler.js:151]: Missed this before :)

```suggestion
const redirectToSparkApi = (req, res, SPARK_API_BASE_URL) => {
```

File: stats/lib/handler.js [hunk 6]
```diff
@@ -102,11 +105,11 @@ const handler = async (req, res, pgPools) => {
   } else if (req.method === 'GET' && url === '/miners/retrieval-success-rate/summary') {
     await respond(fetchMinersRSRSummary)
   } else if (req.method === 'GET' && segs[0] === 'miner' && segs[1] && segs[2] === 'deals' && segs[3] === 'eligible' && segs[4] === 'summary') {
-    await getRetrievableDealsForMiner(req, res, pgPools.api, segs[1])
+    await redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```
Comment (juliangruber) [inline stats/lib/handler.js:108]:
```suggestion
    redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```

File: stats/lib/handler.js [hunk 7]
```diff
@@ -102,11 +105,11 @@ const handler = async (req, res, pgPools) => {
...
   } else if (req.method === 'GET' && segs[0] === 'client' && segs[1] && segs[2] === 'deals' && segs[3] === 'eligible' && segs[4] === 'summary') {
-    await getRetrievableDealsForClient(req, res, pgPools.api, segs[1])
+    await redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```
Comment (juliangruber) [inline stats/lib/handler.js:110]:
```suggestion
    redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```

File: stats/lib/handler.js [hunk 8]
```diff
@@ -102,11 +105,11 @@ const handler = async (req, res, pgPools) => {
...
   } else if (req.method === 'GET' && segs[0] === 'allocator' && segs[1] && segs[2] === 'deals' && segs[3] === 'eligible' && segs[4] === 'summary') {
-    await getRetrievableDealsForAllocator(req, res, pgPools.api, segs[1])
+    await redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```
Comment (juliangruber) [inline stats/lib/handler.js:112]:
```suggestion
    redirectToSparkApi(req, res, SPARK_API_BASE_URL)
```

**Review by juliangruber — CHANGES_REQUESTED** [review]

**Comment by bajtos** [comment]
> > > @bajtos there's a type error
> > 
> > 
> > Do you know why we don't run `npm test:types` as part of `npm test`? Was it an intentional decision? If not, then I'd like to fix that.
> 
> Yes please add that 👌

https://github.com/filecoin-station/spark-stats/pull/208

**Review by juliangruber — APPROVED** [review]
👏

**Pull request merged**